text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+## Compatibility Strategy
+
+The `@embroider/compat` package exists to compile V1 packages to V2. This allows `@embroider/core` to always assume V2 packages as input, so we don't need to wait until every addon is natively available in V2 before we start getting the benefits of Embroider. However, there is still an incentive to convert as many addons as possible to V2, because they build faster and they will be more stable (the v1-to-v2 compilation isn't flawless, we need heuristics and package-specific rules to deal with some dynamic addon behaviors).
+
+It also needs to be possible for an addon published as V2 to work in existing apps on existing ember-cli versions. This is enabled by:
+
+- `ember-auto-import`, which serves as a high-fidelity polyfill for importing directly from NPM. V2 addons natively support importing from NPM, but they should depend on `ember-auto-import` so those imports will have the same meaning when used in classic ember-cli.
+- `@ember/macros`, which shall provide correct semantics regardless of whether the package using them is published as V1 or V2 and regardless of whether the build is being done by classic ember-cli or Embroider. Native V2 packages under Embroider can alway use macros, without an explicit dependency on `@ember/macros`, but they should include the dependency so that macros will work in classic ember-cli.
+- ember-cli already supports a `main` property in **Ember package metadata** and has supported it for many versions. This allows an addon to put its classic `index.js` file in a place other than the package's true `main`. This means that V2 addons can have their runtime `index.js` as `main`, and should point `ember-addon.main` to a `classic.js` file. The `classic.js` file should `require` and `export` a compatibility shim library that we will provide. The compatibility shim will have the classic methods like `treeForAddon`, `treeForPublic` that take the V2-formatted features and present them in a way that classic ember-cli will understand. Since V2 packages are much more static than V1 packages, this shim is expected to not be very complicated.
+
+# How we Teach This
+
+This RFC should have no direct impact on what app authors need to learn. They keep using addons the same way they always have. Future RFCs that take Embroider mainstream _will_ have impact, but that can be discussed then.
+
+The impact on addon authors is more significant. This design is fully backward compatible, and the intention is that all existing addons continue to work (some with worse compatibility hacks than others in the v1-to-v2 compiler). But there will be a demand for addons published in v2 format, since it is expected to result in faster build times. My prediction is that people who are motivated to get their own app build times down will send a lot of PRs to the addons they’re using.
```

Comment:
You mention that the success of getting this format adopted throughout the community will be down to maintainers migrating their addons or other individuals putting together PRs, should there be more emphasis on how we plan to teach and empower people to do this? @jenweber brings up some good points in https://github.com/emberjs/rfcs/pull/507/files#r296484559. 

Will there be guides, migration tutorials, quest issues? Could there be a codemod which gets addons 80% of the way there (if not 100% of the way in some cases)?

Reply:
Yes, this is definitely the highest priority once the addon is merged. In response to your feedback I added a section to the "How we teach" section that says explicitly that we should release an official package with the tooling for addon authors and that package's documentation should serve as the starting point for teaching them.